src/mod/l4_thruster.c

```diff
@@ -332,55 +416,117 @@ void ParseReadRequest(uint8_t* received_buffer,int len){
+			// It will trigger infinite loop if thruser reply with length of message that is greaater then 5 and less then enough to cover 101 registers !
+			//sprintf(print_str, "\nParse Iteration = %d length_of_next_register = %d \n", i,length_of_next_register);
+			//len_print = strlen(print_str);
+			//deb_print_pure_debug((uint8_t *)print_str, len_print);
+
+			//sprintf(print_str, "\n Next RI= %d Next_RI_len=%d  Conversion_multiplier = %.2f \n",next_register_index,length_of_next_register,multiplier  );
+			//len_print = strlen(print_str);
+			//deb_print_pure_debug((uint8_t *)print_str, len_print);
+			current_time = (uint32_t)timGetSystime();
+
+			if((current_time - parse_start_time)> 100){
+				/*
+				 *
+				 * This is exit condition to prevent infinite loop.
+				 * In case thruster replies less bytes then expected - index will never reach exit condition.
+				 * To prevent this we introduce break after 100ms delay
+				 *
+				 * This is not the most elegant solution. But it is assumed that ReadAllRegisters will request amount of bytes
+				 * enough to cover all REGISTER_DATA. Thus in normal operation proccess time exceeded limit warning should never occure
+				 *
+				 */
+
+				sprintf(print_str, "\nParse WARNING: Process time exceeded limit   next_register_index = %d  \n",next_register_index  );
+				len_print = strlen(print_str);
+				deb_print_pure_debug((uint8_t *)print_str, len_print);
+				current_time = (uint32_t)timGetSystime();
+
+
+
+				break;
+			}
 
-		for(int i=0;i<uint16_payload_length;i++){
 
-			//printf("\n i=%d",i);
 			if(length_of_next_register ==1){
 
 				//printf("\n-1------------------Next Register Index = %d",next_register_index);
 				VALUE_UINT8 = received_data[i];
 				multiplier = CONVERSION_DOUBLE[next_register_index];
 				ACTUAL_VALUE = (double)VALUE_UINT8 / multiplier;
 				REGISTER_DATA[next_register_index]=ACTUAL_VALUE;
+				//next_register_index = next_register_index+1;
+
+				//sprintf(print_str, "\n [%d] ACTUAL_VALUE= %.6f Conversion=%.6f  uint8_value =%d  \n",next_register_index,ACTUAL_VALUE,multiplier,VALUE_UINT8  );
+				//len_print = strlen(print_str);
+				//deb_print_pure_debug((uint8_t *)print_str, len_print);
+
 				next_register_index = next_register_index+1;
-				//printf("\nActual Value =%f \n",ACTUAL_VALUE);
-				//printf("\n multiplier =%f \n",multiplier);
-				//printf("\n INTEGER Value =0x%00x\n",VALUE_UINT8);
+				i = i+1;
 
 			}
 
 			if(length_of_next_register==2){
 
 				//printf("\n 2------------Next Register Index = %d \n",next_register_index);
-				VALUE_UINT16 = (received_data[i+1] <<8 )| received_data[i];
+				VALUE_UINT16 = (received_data[i+1] <<8 )| received_data[i]; // here problem / Compiles value out of wrong bytes
 				double multiplier = CONVERSION_DOUBLE[next_register_index];
 				ACTUAL_VALUE = (double)VALUE_UINT16 / multiplier;
 				REGISTER_DATA[next_register_index]=ACTUAL_VALUE;
+
```

Comment:
I want to leave debug messages commented instead of deleting it (in case I ever need to investigate into parser again)

For now parser works as expected.